src/python/pants/jvm/util_rules.py

```diff
@@ -18,23 +25,25 @@ class ExtractFileDigest:
 
 @rule
 async def digest_to_file_digest(request: ExtractFileDigest) -> FileDigest:
-    """TODO(#11806): This is just a workaround; this extraction should be provided directly by the engine."""
-
     digest = await Get(Digest, DigestSubset(request.digest, PathGlobs([request.file_path])))
-    digest_contents = await Get(DigestContents, Digest, digest)
-    if len(digest_contents) == 0:
+    digest_entries = await Get(DigestEntries, Digest, digest)
+
+    if len(digest_entries) == 0 or (
+        len(digest_entries) == 1 and digest_entries[0] == Directory("")
```

Comment:
Maybe just filter out `Directory` instances and then no need to special case? That is:

```
digest_entries = [entry for entry in digest_entries if isinstance(entry, FileEntry)]
```

Alternatively, maybe `DigestEntries` needs to return an empty collection when there are no contents instead of stating that there is an empty root directory? (Which would points to fixing the `DigestEntries` intrinsic in a separate PR)

I'd say the latter is probably the solution. Thoughts?

Reply:
I'll go with the first for this PR (for ease of merging) and then follow up with the latter.